### PR TITLE
[drape] Optimize Spline direction/length calculations.

### PR DESCRIPTION
**Pull request by vng**
- One more case to optimize Spline::IsProlonging when simplify mode (z10-12). Need refactor ClipSpline routine for separate Spline and SplineEx cases.
- Make simplification in shape coordinates (LineShape::Construct) where we calculate normals and tangents?

Done: If the tile picture is good, can keep only the second commit and calculate reusable tangent/length always.

https://github.com/organicmaps/organicmaps/issues/6137
Closes https://github.com/organicmaps/organicmaps/issues/3295
Closes https://github.com/organicmaps/organicmaps/issues/1734
Closes https://github.com/organicmaps/organicmaps/issues/4702

### CODE REVIEW

File: drape_frontend/apply_feature_functors.cpp [hunk 1]
```diff
@@ -777,7 +777,7 @@ void ApplyLineFeatureGeometry::operator() (m2::PointD const & point)
   {
     if (m_simplify &&
         ((m_spline->GetSize() > 1 && point.SquaredLength(m_lastAddedPoint) < m_minSegmentSqrLength) ||
-        m_spline->IsPrelonging(point)))
+          m_spline->IsProlonging(point)))
```
Comment (biodranik) [inline drape_frontend/apply_feature_functors.cpp:780]: That looks worse than another approach:
```
    if (m_simplify
        && ((m_spline->GetSize() > 1 && point.SquaredLength(m_lastAddedPoint) < m_minSegmentSqrLength)
        || m_spline->IsProlonging(point)))
```
Agree?

File: geometry/spline.cpp [hunk 2]
```diff
@@ -52,20 +76,19 @@ void Spline::ReplacePoint(PointD const & pt)
...
   PointD dir = pt - m_position.back();
   if (dir.IsAlmostZero())
     return true;
-
   dir = dir.Normalize();
-  PointD prevDir = m_direction.back().Normalize();
 
-  double const MAX_ANGLE_THRESHOLD = 0.995;
-  return std::fabs(DotProduct(prevDir, dir)) > MAX_ANGLE_THRESHOLD;
+  PointD const prevDir = (m_position[sz-1] - m_position[sz-2]).Normalize();
```
Comment (biodranik) [inline geometry/spline.cpp:90]:
```suggestion
  PointD const prevDir = (m_position[sz - 1] - m_position[sz - 2]).Normalize();
```

File: geometry/spline.cpp [hunk 3]
```diff
@@ -52,20 +76,19 @@ void Spline::ReplacePoint(PointD const & pt)
...
     return false;
 
   PointD dir = pt - m_position.back();
   if (dir.IsAlmostZero())
     return true;
-
   dir = dir.Normalize();
-  PointD prevDir = m_direction.back().Normalize();
 
-  double const MAX_ANGLE_THRESHOLD = 0.995;
```
Comment (biodranik) [inline geometry/spline.cpp:67]: Let's leave `double constexpr kMaxAngleThreshold = 0.995;` for readability.

File: geometry/spline.hpp [hunk 4]
```diff
@@ -77,14 +76,25 @@ class Spline
...
+  void InitDirections() const;
 
   std::vector<PointD> m_position;
-  std::vector<PointD> m_direction;
-  std::vector<double> m_length;
+  mutable std::vector<PointD> m_direction;
+  mutable std::vector<double> m_length;
+};
+
+class SplineEx : public Spline
```
Comment (biodranik) [inline geometry/spline.hpp:89]: Maybe EditableSpline? Or AddReplaceSpline?

**Review by biodranik — APPROVED** [review]

File: drape_frontend/apply_feature_functors.cpp (hunk 1 above)
Comment (vng) [reply drape_frontend/apply_feature_functors.cpp:780]: I don't like it, and this is confusing: && and II here are not in one level.

File: geometry/spline.cpp (hunk 3 above)
Comment (vng) [reply geometry/spline.cpp:67]: This constant name is confusing. This is vector-dot-product threshold, which becomes clear one line below without constant.

**Review by biodranik — APPROVED** [review]

File: drape_frontend/apply_feature_functors.cpp (hunk 1 above)
Comment (vng) [reply drape_frontend/apply_feature_functors.cpp:780]: Then something like this:
```
if (m_simplify
        && ((m_spline->GetSize() > 1 && point.SquaredLength(m_lastAddedPoint) < m_minSegmentSqrLength)
                || m_spline->IsProlonging(point)))
```
I don't like it either :)

**Comment by pastk** [comment]
> * One more case to optimize Spline::IsProlonging when simplify mode (z10-12).

Why there is this extra run-time z10-12 simplification at all? Is it to simplify user tracks?

**Review by biodranik — APPROVED** [review]

File: geometry/spline.cpp [hunk 5]
```diff
@@ -52,20 +60,20 @@ void Spline::ReplacePoint(PointD const & pt)
   AddPoint(pt);
 }
 
-bool Spline::IsPrelonging(PointD const & pt)
+bool Spline::IsProlonging(PointD const & pt) const
```
Comment (pastk) [inline geometry/spline.cpp:63]: its used in apply_feature_functors only and contains similar min length condition which is redundant;

I think makes sense to do both length and angle comparisons in one place only - either apply_feature_functors or here

File: drape_frontend/line_shape.cpp [hunk 6]
```diff
@@ -352,14 +347,47 @@ void LineShape::Construct(TBuilder & builder) const
...
+template <class FnT> void LineShape::ForEachSplineSection(FnT && fn) const
+{
+  std::vector<m2::PointD> const & path = m_spline->GetPath();
+  ASSERT(!path.empty(), ());
+  size_t const sz = path.size() - 1;
+
+  for (size_t i = 0, j = 1; j <= sz; ++j)
+  {
+    /// @todo Make this kind of fitration in Spline?
+    if (path[i].EqualDxDy(path[j], kMwmPointAccuracy) && j < sz)
```
Comment (pastk) [inline drape_frontend/line_shape.cpp:359]: when is this possible? for user tracks only? looks like its almost duplicating the filtering in apply_feature_functors

**Review by pastk — COMMENTED** [review]
If this kind of filtering is needed for user tracks only then we should disable it for other lines to avoid useless work.

Also filtering should be done once only in apply_feature_functors to avoid re-filtering in line_shape and re-calculating tangents and lengths.

(and ideally we should store/cache simplified user tracks somewhere to avoid re-filtering every time we draw them)

**Comment by pastk** [comment]
Also what was the original bug exactly? (and the fix)
I couldn't spot it amongst the refactoring...

**Comment by pastk** [comment]
Also its not clear for me how is it happening for a border line #4702? Its supposed to be simplified already in geom3 (as opposed to roads which are not simplified on geom3).

So it might mean that the thresholds used for the in-generator simplification and run-time simplification are not synced properly.

**Comment by vng** [comment]
As discussed:
- ```if (path[i].EqualDxDy(path[j], kMwmPointAccuracy)``` happens. Making correct traversal along a path is a fix.
- Obviously, generator simplification criteria and kMwmPointAccuracy have _different nature_ and should be revised.
ToDo https://github.com/organicmaps/organicmaps/issues/6137#issuecomment-1737294455
- IsProlonging is one more crutch to reduce number of vertices on z10-12 that _should not_ influence on visual tile quality
ToDo: Revise/unify this logic.

We can make _the best_ generator simplification, but anyway some runtime filtering should be present. At least because of different target device's DPI. Look at this :)
```m_minSegmentSqrLength(base::Pow2(4.0 * df::VisualParams::Instance().GetVisualScale() / currentScaleGtoP))```

**Pull request merged**